### ISSUE TITLE
hangs indefinitely at os.waitpid()

### ISSUE THREAD

**Issue by hudsontek** [pytorch/pytorch#48382]
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description

When using multiprocessing to create a network, the program hangs indefinitely. Interrupting the program will show that the program hangs at os.waitpid

## Code example

import torch
from torch import nn
import torch.multiprocessing as mp

def normalized_columns_initializer(weights, std=1.0):
    out = torch.randn(weights.size())
    out *= std / torch.sqrt(out.pow(2).sum(1, keepdim=True))
    return out

class ACNet(nn.Module):
    def __init__(self, max_addrs, num_loc, hidden_dim = 64):
        super().__init__()
        self.loc_linear = nn.Linear(hidden_dim, num_loc) # select insert location

        print('start init')
        self.loc_linear.weight.data = normalized_columns_initializer(
            self.loc_linear.weight.data, 0.01)    #todo: this line causes deadlock somehow
        self.loc_linear.bias.data.fill_(0)

        print("init....")

def create_model(dim):
    mdl = ACNet(100, dim)
    print('model is created')

if __name__ == '__main__':
    shared_model = ACNet(100, 512)
    p = mp.Process(target=create_model, args=(512,))
    p.start()
    p.join()

## System Info

PyTorch version: 1.7.0+cpu
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Linux Mint 19.1 Tessa (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 7.0.0 (tags/RELEASE_700/final)
CMake version: version 3.12.3

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] torch==1.7.0+cpu
[pip3] torchaudio==0.7.0
[pip3] torchvision==0.8.1+cpu
[conda] Could not collect

**Comment by mruberry** [comment]
Thank you for reporting this issue, @hudsontek. Unfortunately I was unable to reproduce it on Colab. Maybe this is a local issue only?

**Comment by hudsontek** [comment]
I don't think so. I have successfully reproduce it on other machines. Please do use Linux. This issue is not reproducible on Windows, as far as I know. So it may not be reproducible in my guess. So please try it on Linux.

**Comment by mruberry** [comment]
You're right, this reproduces on my unix machine. Commenting out the line creating the model before calling `mp.Process` resolves the issue. https://pytorch.org/docs/master/notes/multiprocessing.html doesn't seem to describe this behavior, although it suggests using distributed data parallel instead of multiprocessing.

**Comment by ezyang** [comment]
I bet this is because you're initializing threads prior to forking, which leaves the forked thread in a bad state. The size matters because we don't initialize OpenMP unless the input is large enough. Try adding

```
    mp.set_start_method('spawn')
```

to the start of your program

**Comment by ezyang** [comment]
I wouldn't close this issue because we should be checking if OpenMP was initialized after a fork and giving an error message in this case.

**Comment by zou3519** [comment]
@mruberry says: our docs aren't clear about what is going on here

@ezyang there's a simple docs update we can do here, but really we can also give better error messages here.

@mruberry @ezyang says: we should do both. 

We may already have an issue about this. @albanD says #17199 is this same issue for openmp. There's another one for autograd but it was fixed/closed.

**Comment by hudsontek** [comment]
> I bet this is because you're initializing threads prior to forking, which leaves the forked thread in a bad state. The size matters because we don't initialize OpenMP unless the input is large enough. Try adding
> 
> ```
>     mp.set_start_method('spawn')
> ```
> 
> to the start of your program

This does solve the problem. But still, this is a weird bug. Different behiviors on different platforms.

**Comment by ShiningDeadFishEyes** [comment]
> I bet this is because you're initializing threads prior to forking, which leaves the forked thread in a bad state. The size matters because we don't initialize OpenMP unless the input is large enough. Try adding
> 
> ```
>     mp.set_start_method('spawn')
> ```
> 
> to the start of your program

It worked, tks!
